### ISSUE TITLE
Rule Change: when using no-restricted-imports, be able to have importNames as regexp

### ISSUE THREAD

**Issue by epaillous** [eslint/eslint#17303]
### What rule do you want to change?

no-restricted-imports

### What change to do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new option

### Example code

```js
{
      "files": ["app/javascript/components/**/*"],
      "rules": {
        "no-restricted-imports": ["error", {
          "patterns": [{
            "group": ["queries/*"],
            "importNames": ["^use"],
          }]
        }]
      }
    }
```

### What does the rule currently do for this code?

For now it's not possible to have regexps inside `importNames`. In my case I want to forbid importing elements starting by "use" from one module.

### What will the rule do after it's changed?

Being able to restrict some imports matching a regexp pattern. 

### Participation

- [X] I am willing to submit a pull request to implement this change.

### Additional comments

I've started to try to implement something, but I am thinking maybe we should use a dedicated property such as importPatterns to avoid confusion with exact names and regexp. What do you think ?

Thanks for this awesome library!

**Comment by Rec0iL99** [comment]
Hi @epaillous, thanks for the issue. If accepted, I think we would have to follow the [gitignore specification](https://git-scm.com/docs/gitignore) because that is used in `patterns`.

**Comment by nzakas** [comment]
This seems like a logical extension to the existing rule given that you can already specify filename patterns to exclude, but I'll defer to @mdjermanovic for feedback on the best way to implement such a feature.

**Comment by nzakas** [comment]
@mdjermanovic thoughts on this?

**Comment by Rec0iL99** [comment]
@mdjermanovic friendly ping. Waiting for your feedback.

**Comment by mdjermanovic** [comment]
Makes sense to me.

> I've started to try to implement something, but I am thinking maybe we should use a dedicated property such as importPatterns to avoid confusion with exact names and regexp. What do you think ?

I agree this should be a new property to avoid confusion with exact names.

Since the term "pattern" is used in this rule for gitignore patterns, it might be better to use "regex" for regex patterns.

Also, I think we can simplify the schema by allowing only 1 pattern, as the functionality of multiple patterns can be achieved by using `|` in a single pattern.

So, the new property could be `importNamesRegex: string`.

> ```js
>"no-restricted-imports": ["error", {
>    "patterns": [{
>        "group": ["queries/*"],
>            "importNamesRegex": "^use"
>        }]
>}]
> ```

Thoughts?

**Comment by nzakas** [comment]
> Since the term "pattern" is used in this rule for gitignore patterns, it might be better to use "regex" for regex patterns.

It seems like we use "pattern" in most rules for this purpose, so to be consistent across rules, I think `importNamePatterns` make sense and is distinctive enough from `patterns` that I don't think it will cause confusion.

**Comment by nzakas** [comment]
@mdjermanovic thoughts on my last comment?

**Comment by mdjermanovic** [comment]
> It seems like we use "pattern" in most rules for this purpose, so to be consistent across rules, I think `importNamePatterns` make sense and is distinctive enough from `patterns` that I don't think it will cause confusion.

Agreed on "pattern".

However, I still think there's no need for multiple patterns, so `importNamePattern: string`?

**Comment by nzakas** [comment]
> However, I still think there's no need for multiple patterns, so importNamePattern: string?

Works for me. 👍

**Comment by nzakas** [comment]
@epaillous sorry this took so long to resolve. Are you still interested in submitting a PR?

**Comment by Tanujkanti4441** [comment]
if no replies can i take this.

**Comment by nzakas** [comment]
@Tanujkanti4441 go for it!
